OPHD/UI/StringTable.cpp

```diff
@@ -0,0 +1,221 @@
+{
+	return cells[getCellIndex(coordinate)];
+}
+
+StringTable::Cell& StringTable::at(std::size_t column, std::size_t row)
+{
+	return cells[getCellIndex(CellCoordinate(column, row))];
+}
+
+StringTable::StringTable(std::size_t columns, std::size_t rows) : columnCount(columns), rowCount(rows)
+{
+	cells.resize(columns * rows);
+
+	defaultFont = NAS2D::Utility<FontManager>::get().font(constants::FONT_PRIMARY, constants::FONT_PRIMARY_NORMAL);
+	defaultTitleFont = NAS2D::Utility<FontManager>::get().font(constants::FONT_PRIMARY_BOLD, constants::FONT_PRIMARY_NORMAL);
+}
+
+void StringTable::draw(NAS2D::Renderer& renderer) const
+{
+	for (std::size_t i = 0; i < cells.size(); ++i)
+	{
+		const auto& cell = cells.at(i);
+
+		NAS2D::Color textColor = cell.textColor != Cell::ColorEmpty ? cell.textColor : defaultTextColor;
+
+		renderer.drawText(*getCellFont(i), cell.text, position + cell.textRelativePosition, textColor);
+	}
+}
+
+void StringTable::setPosition(NAS2D::Point<float> tablePosition)
+{
+	this->position = tablePosition;
+}
+
+NAS2D::Point<float> StringTable::getPosition() const
+{
+	return position;
+}
+
+void StringTable::setDefaultFont(NAS2D::Font& font)
+{
+	defaultFont = &font;
+}
+
+void StringTable::setDefaultTitleFont(NAS2D::Font* font)
+{
+	defaultTitleFont = font;
+}
+
+void StringTable::setDefaultTextColor(NAS2D::Color color)
+{
+	defaultTextColor = color;
+}
+
+void StringTable::setHorizontalPadding(float padding)
+{
+	horizontalPadding = padding;
+}
+
+void StringTable::setVerticalPadding(float padding)
+{
+	verticalPadding = padding;
+}
+
+void StringTable::setColumnJustification(std::size_t column, Justification justification)
+{
+	for (std::size_t row = 0; row < rowCount; ++row)
+	{
+		this->at(column, row).justification = justification;
+	}
+}
+
+void StringTable::computeRelativeCellPositions()
+{
+	// Must be at least 1 row and 1 column to compute position values
+	if (columnCount == 0 || rowCount == 0)
+	{
+		return;
+	}
+
+	auto columnWidths = computeColumnWidths();
+	auto rowHeights = computeRowHeights();
+
+	float columnOffset = 0;
+	for (std::size_t column = 0; column < columnCount; ++column)
+	{
+		float rowOffset = 0;
+		for (std::size_t row = 0; row < rowCount; ++row)
+		{
+			auto cellIndex = getCellIndex(CellCoordinate(column, row));
+			cells[cellIndex].textRelativePosition = { columnOffset, rowOffset };
+			accountForCellJustification(cellIndex, columnWidths[column]);
+
+			rowOffset += rowHeights[row] + verticalPadding;
+		}
+
+		columnOffset += columnWidths[column] + horizontalPadding;
+	}
+}
+
+void StringTable::accountForCellJustification(std::size_t index, float columnWidth)
+{
+	auto& cell = cells[index];
+
+	switch (cell.justification)
+	{
+	case (Justification::Left):
+		return; // No modification required for left justifited
+	case (Justification::Right):
+		cell.textRelativePosition.x += columnWidth - getCellFont(index)->width(cell.text);
```

Comment:
Correct, since the cell purposely does not know what the StringTable's default font is, the cell's font should not be directly queried, but StringTable::getCellFont should be called instead. See comment above that we could just initialize the default on construction of StringTable if this behaviour is not appreciated.